File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-3.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for t in tokens:
            try:
                stack.append(int(t))
            except ValueError:
                v2 = stack.pop()
                v1 = stack.pop()
                if t == '+':
                    stack.append(v1+v2)
                elif t == '-':
                    stack.append(v1-v2)
                elif t == '*':
                    stack.append(v1*v2)
                elif t == '/':
                    stack.append(int(v1/v2))
        return stack[-1]

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-3.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> st;
        int i = 0;
        int ret = 0;
        for (i=0; i<heights.size(); i++) {
            while (!st.empty() && heights[st.top()]>=heights[i]) {
                int height = heights[st.top()]; st.pop();
                int width = i;
                if (!st.empty()) width -= st.top()+1;
                ret = max(ret, width*height);
            }
            st.push(i);
        }
        while (!st.empty()) {
            int height = heights[st.top()]; st.pop();
            int width = heights.size();
            if (!st.empty()) width -= st.top()+1;
            ret = max(ret, width*height);
        }
        return ret;
    }
};
